src/util/nextThought.js

```diff
@@ -0,0 +1,252 @@
+  const matchedContextWithRanks = contextWithThoughtRank(getMatchedContext(context, contextChain))
+  return [...matchedContextWithRanks, context, ...head(contextChain).slice(1)]
+}
+
+/**
+ * Returns the context of the current thought by stripping out any context views
+ * @returns {array} context
+ */
+const getContextFromContextChain = contextChain => {
+  // last of second last item in context chain gives us the current context
+  const context = head(head(contextOf(contextChain)))
+  const matchedContext = getMatchedContext(context, contextChain)
+  return [...matchedContext.context, context.value, ...head(contextChain).slice(1).map(context => context.value)]
+}
+
+/**
+ * Computes the value of next thoughts  and context chain if-
+ * 1) the focused thought has context view open
+ * 2) the focus is on a context
+ * Delegates control to nextInThoughtView if none of the above conditions meet
+ * @param {string} value - The value of focused thought
+ * @param {string} rank - Rank of focused thought
+ * @param {array} path - Path to focsued thought
+ * @param {array} rankedContext - Context with rank
+ * @param {array} contextChain - ContextChain for the focused thought
+ * @param {(boolean | undefined)} ignoreChildren - Used to ignore the children context if they've been traversed already
+*/
+const nextInContextView = (value, rank, path, rankedContext, contextChain, ignoreChildren) => {
+  const { showHiddenThoughts, thoughtIndex } = store.getState()
+
+  if (rankedContext.length === 0 || path.length === 0) return null
+
+  const context = pathToContext(rankedContext)
+  const firstChild = perma(() => firstChildOfContextView(path, thoughtIndex))
+
+  const contextWithoutChildren = isContextViewActive(pathToContext(path)) &&
+    getContexts(head(path).value).length < (ALLOW_SINGLE_CONTEXT ? 2 : 1)
+
+  if (contextWithoutChildren && contextChain.length === 1) {
+    return nextInThoughtView(value, context, rank, path, contextChain, true)
+  }
+
+  // if the focus is on a thought with context view open, move it into context view - jump in
+  if (!contextWithoutChildren && !ignoreChildren && isContextViewActive(pathToContext(path)) && firstChild()) {
+    const currentThought = head(path)
+    // jump out if there are no context children
+    return {
+      nextThoughts: [...firstChild(), currentThought],
+      contextChain
+    }
+  }
+  // if the focus is on or within a context
+  else if (isContextViewActive(pathToContext(rankedContext))) {
+    const firstChild = perma(() => firstChildOfThoughtView(getContextFromContextChain(contextChain) || RANKED_ROOT, showHiddenThoughts))
+
+    const nextSibling = nextSiblingContext(rank, context, thoughtIndex, showHiddenThoughts)
+    const rankedContextHead = head(rankedContext)
+
+    return !ignoreChildren && firstChild() ? {
+      nextThoughts: unroot([...path, firstChild()]),
+      contextChain: []
+    }
+      : nextSibling ? {
+        nextThoughts: [...nextSibling, rankedContextHead],
```

Comment:
I'm going to start here, where `nextSibling` is used, and work my way backwards.

`nextSibling` is the result of calling `nextSiblingContext`, and it is the only place `nextSiblingContext` is used, so we don't have the added complexity of multiple cases.

[`nextSiblingContext` calls `contextWiththoughtRank`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L61), which is the function that laboriously converts a `Context` to a `Path`. (`contextWithThoughtRank` is used in other places, but for now I am only considering it as it is used in `nextSiblingContext`.)

The problem is that `contextWithThoughtRank` goes through all the trouble of recreating ranks, when we already have all of the ranks, not including the next sibling, in `rankedContext` [that is provided to `nextInContextView`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L131). Recreating the rank at every level is unnecessary. We just need to concatenate the correct `Child` onto `rankedContext`.

This is the kind of optimization I'm taking about, and may apply to the other uses of `contextWithThoughtRank`. It's usually just the current level where we may need to get a thought or rank that we don't have. All of the ancestors are being passed around in `path` and `contextRanked`.

Reply:
In this particular case that you've pointed out, we're not computing the ranks of the entire path. As you mentioned, [`nextSiblingContext calls contextWithThoughRank`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L61), the ranks are only computed for the **context** that's being passed, which is the `nextSibling` and its rank information is not included in the `rankedContext`.
The only case where we're computing the `ranks` for the full path is when we need to compute the [`nextUncleInThoughtView` within a context view](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L194). Reason for the same is mentioned [here](https://github.com/cybersemics/em/pull/535#discussion_r421526834)

Reply:
> In this particular case that you've pointed out, we're not computing the ranks of the entire path. As you mentioned, [`nextSiblingContext calls contextWithThoughRank`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L61), the ranks are only computed for the **context** that's being passed, which is the `nextSibling` and its rank information is not included in the `rankedContext`.

The rank of the `head` of the next sibling does need to be calculated. I agree with that. But the ranks of the rest of the sibling context should already be known. After all, it is the sibling, so it has the same ancestors as `path`. It seems like we only need to calculate one rank, the sibling, not the rank for every thought in the context. We just need to copy them over from `path`. If this is the case, It seems unnecessary to [iterate through `contextInfo.context`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L38).

Here is the case I am considering. Set the cursor on `a.b.c.d.x~` and activate `cursorDown`. You'll see that [`matchedContext`](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L40-L44) gets called four times (for each ancestor of `x`).

```
- a
  - b
    - c
      - d
        - x
      - e
        - x
```

> The only case where we're computing the `ranks` for the full path is when we need to compute the [`nextUncleInThoughtView` within a context view](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L194). Reason for the same is mentioned [here](https://github.com/cybersemics/em/pull/535#discussion_r421526834)

I'd like to confirm the exact case you're talking about, but if [`getPathFromContextChain` is called](https://github.com/cybersemics/em/blob/bbd219fb54273a0dd15a65b4f247f854733d29f9/src/util/nextThought.js#L194), then `contextWithThoughtRank` will be called and the same redundant calls to `matchedContext` will occur.

Perhaps I am missing something, but more than happy to discuss either way!